xcm/pallet-xcm/src/tests.rs

```diff
@@ -0,0 +1,180 @@
+	let balances =
+		vec![(ALICE, INITIAL_BALANCE), (ParaId::from(PARA_ID).into_account(), INITIAL_BALANCE)];
+	new_test_ext_with_balances(balances).execute_with(|| {
+		let amount = 10 * ExistentialDeposit::get();
+		let weight = 2 * BaseXcmWeight::get();
+		assert_eq!(Balances::total_balance(&ALICE), INITIAL_BALANCE);
+		assert_ok!(XcmPallet::teleport_assets(
+			Origin::signed(ALICE),
+			RelayLocation::get(),
+			RelayLocation::get(),
+			vec![ConcreteFungible { id: MultiLocation::Null, amount }],
+			weight,
+		));
+		assert_eq!(Balances::total_balance(&ALICE), INITIAL_BALANCE - amount);
+		assert_eq!(
+			last_event(),
+			Event::XcmPallet(crate::Event::Attempted(Outcome::Complete(weight)))
+		);
+	});
+}
+
+#[test]
+fn reserve_transfer_assets_works() {
+	let balances =
+		vec![(ALICE, INITIAL_BALANCE), (ParaId::from(PARA_ID).into_account(), INITIAL_BALANCE)];
+	new_test_ext_with_balances(balances).execute_with(|| {
+		let amount = 10 * ExistentialDeposit::get();
+		let weight = BaseXcmWeight::get();
+		let dest: MultiLocation =
+			Junction::AccountId32 { network: NetworkId::Any, id: ALICE.into() }.into();
+		assert_eq!(Balances::total_balance(&ALICE), INITIAL_BALANCE);
+		assert_ok!(XcmPallet::reserve_transfer_assets(
+			Origin::signed(ALICE),
+			Parachain(PARA_ID).into(),
+			dest.clone(),
+			vec![ConcreteFungible { id: MultiLocation::Null, amount }],
+			weight
+		));
+		// Alice spent amount
+		assert_eq!(Balances::free_balance(ALICE), INITIAL_BALANCE - amount);
+		// Destination account (parachain account) has amount
+		let para_acc: AccountId = ParaId::from(PARA_ID).into_account();
+		assert_eq!(Balances::free_balance(para_acc), INITIAL_BALANCE + amount);
+		assert_eq!(
+			sent_xcm(),
+			vec![(
+				Parachain(PARA_ID).into(),
+				Xcm::ReserveAssetDeposit {
+					assets: vec![ConcreteFungible { id: Parent.into(), amount }],
+					effects: vec![buy_execution(weight), DepositAsset { assets: vec![All], dest },]
+				}
+			)]
+		);
+		assert_eq!(
+			last_event(),
+			Event::XcmPallet(crate::Event::Attempted(Outcome::Complete(weight)))
+		);
+	});
+}
+
+#[test]
+fn execute_withdraw_to_teleport_works() {
+	let balances =
+		vec![(ALICE, INITIAL_BALANCE), (ParaId::from(PARA_ID).into_account(), INITIAL_BALANCE)];
+	new_test_ext_with_balances(balances).execute_with(|| {
+		let amount = 10 * ExistentialDeposit::get();
+		let weight = 3 * BaseXcmWeight::get();
+		let dest: MultiLocation =
+			Junction::AccountId32 { network: NetworkId::Any, id: ALICE.into() }.into();
+		let teleport_effects = vec![
+			buy_execution(0),
+			Order::DepositAsset { assets: vec![All], dest: Parachain(PARA_ID).into() },
+		];
+		assert_eq!(Balances::total_balance(&ALICE), INITIAL_BALANCE);
+		assert_ok!(XcmPallet::execute(
+			Origin::signed(ALICE),
+			Box::new(Xcm::WithdrawAsset {
+				assets: vec![ConcreteFungible { id: MultiLocation::Null, amount }],
+				effects: vec![
+					buy_execution(weight),
+					Order::InitiateTeleport {
```

Comment:
If `Teleport` here is arbitrary, might as well test `Withdraw` -> `Deposit` and check that another user got the funds or something?